feat(i18n): extract text from workspace elements

shagrirse opened the pull request:
## Problem

Closes #8327 

## Solution

Extracted text from:
- All text in the scope of this issue's components
- `Workspace Header/EditMenu/FormRow`, `SaveSecretKeyScreen`, and `FormResponseOptions` components
- Aria label in `WorkspaceMenuTabs` and `ProductItem.tsx`
- Refactored `create` and `delete` form modals to fit new locale structure

Relevant test and one constant's [1] have been extracted with method outlined by @littlemight in https://github.com/opengovsg/FormSG/pull/8314#issuecomment-2812180569

Saving as a draft first so that I can extract other workspace-relevant texts as well as give @littlemight (?) to review the implementations of text extraction for constants and tests that follow his approach.


**Considerations/Side Notes**

- [1] I've scaffolded the locale files for validation constants, but only focusing on the constant that is used within this issue's scope (`CANNOT_TRANSFER_OWNERSHIP_TO_SELF`), all the others are used throughout the project and I'm holding off on that until the aforementioned approach is deemed to be ok to officially use.

**Breaking Changes** 
- [ ] Yes - this PR contains breaking changes
    - Details ...
- [x] No - this PR is backwards compatible  


## Tests

### Regression
- [ ] WhatsNewDrawer
  - [ ] Open the what's new drawer
  - [ ] Ensure extracted texts are displayed correctly
- [ ] TransferOwnershipModal
  - [ ] Go to `user` > `Transfer All Forms`
  - [ ] Attempt to transfer a form to yourself and a valid email
  - [ ] Ensure all validation errors/texts are properly extracted
- [ ] WorkspaceMenuHeader
  - [ ] Add a new folder
  - [ ] Ensure aria-labels are properly extracted
  - [ ] Ensure sidebar 'Folders' header is properly extracted
- [ ] WorkspaceEditMenu, WorkspaceHeader and create/delete modal
  - [ ] Create a new form
  - [ ] Ensure storage and mrf modes and their respective feature list items are extracted correctly
  - [ ] Try to press `Back` on your browser when the modal prompts you to download your secret key, the confirmation popup text should be properly extracted
  - [ ] Ensure aria labels are properly extracted
  - [ ] Go into workspace, and search for forms (no results should show plural in the workspace header)
  - [ ] Create a new folder
  - [ ] Try to delete/rename the folder
  - [ ] Ensure all texts are properly extracted
- [ ] WorkspaceFormRow
  - [ ] Ensure aria labels are properly extracted
- [ ] WorkspacePage
  - [ ] Ensure default title of workspace side bar is properly extracted (should be 'All forms')
  - [ ] Ensure email mode retirement notification banner/inline message component text is properly extracted
- [ ] WorkspaceModal
  - [ ] Create a workspace (folder), ensure the text (header, footer, buttons) is properly extracted
  - [ ] Delete three workspaces (one with a single form, one with multiple forms, and one with no forms). Ensure text is extracted and plurals are handled.
  - [ ] Rename a workspace, ensure the text (header, footer, buttons) is properly extracted.
- [ ] WorkspaceValidation
  - [ ] When creating/renaming a workspace, go over and under the character limit, add special characters and leave the field as empty. Ensure all validation errors are properly extracted.

shagrirse committed:
feat(i18n): extract text from workspace elements

shagrirse committed:
feat(i18n): extract text - workspace Header/EditMenu/FormRow

- refactor(i18n): create/delete form modals

- feat(i18n): extract aria text from ProductItem.tsx

shagrirse committed:
feat(i18n): extract more text from workspace elements

- change keys that were refactored in DupeFormWizardProvider.tsx and UseTemplateWizardProvider.tsx

- extract remaining text from SaveSecretKeyScreen, WorkspacePage, EmailModeFeedbackAndCreateScreen,  FormResponseOptions components

- Remove simplur import from WorkspaceHeader

littlemight commented:
@shagrirse sorry for idling on this, just came back from my leave.

I'll take a look on this after I worked on my PRs as well

LoneRifle commented:
@shagrirse - apologies, but the codebase has moved on a bit. Would you be happy to work out the merge conflicts?

LoneRifle commented on frontend/src/i18n/locales/constants/validation/en-sg.ts:
Not really happy with the naming conventions in the keys in this file, since it breaks from traditional conventions for i18next. Could we revisit this please?

shagrirse replied:
Apologies! I've amended the naming of the keys to the traditional conventions for i18n :smile:

LoneRifle commented on frontend/src/constants/validation.ts:
```diff
@@ -6,7 +6,7 @@ export const INVALID_DROPDOWN_OPTION_ERROR =
 
 export const CANNOT_TRANSFER_OWNERSHIP_TO_SELF =
-  'You cannot transfer ownership to yourself'
+  'constants.validationConstants.CANNOT_TRANSFER_OWNERSHIP_TO_SELF'
```
I'd generally prefer just removing the constant and quoting the i18next key itself, since that is intrinsically a constant too

shagrirse replied:
Removed the translated constant from this file and quoted the actual key in the component that uses it. Didn't remove the rest as they're still being used/not in the scope of the PR.

LoneRifle commented on frontend/src/features/workspace/components/CreateFormModal/CreateFormModalContent/EmailModeFeedbackAndCreateScreen.tsx:
We're intending to phase out e-mail mode altogether, so we can consider not actually changing this screen

LoneRifle commented on frontend/src/features/workspace/components/CreateFormModal/CreateFormModalContent/SaveSecretKeyScreen.tsx:
```diff
@@ -173,7 +171,9 @@ interface SaveSecretKeyScreenProps {
   useSaveSecretKey = useSaveSecretKeyDefault,
 }: SaveSecretKeyScreenProps): JSX.Element => {
-  const { t } = useTranslation()
+  const { t } = useTranslation('translation', {
+    keyPrefix: 'features.workspace.modals.forms.create',
+  })
```
oh, that's cool! Didn't know this could be done.

LoneRifle commented on frontend/src/i18n/locales/constants/validation/index.ts:
Likewise here concerning key naming convention

LoneRifle reviewed (changes requested):
lgtm otherwise

shagrirse committed:
fix(i18n): implementation adn naming conventions of constants

chore(i18n): revert EmailModeFeedbackAndCreateScreen to origin/develop version

shagrirse commented:
@LoneRifle resolved the merge conflicts and the requested changes. Do let me know if there's anything else :+1:

shagrirse committed:
feat(i18n): extracted text from workspace-related modals

- chore(i18n): remove unused locales relating to email mode

- feat(i18n): extracted text from workspace title validation rules

- refactor(i18n): workspace modal locales

shagrirse commented:
@LoneRifle added a new commit based on the expanded scope of the WorkspaceModal component (create, rename, delete, and validation rules) which was added earlier today. Totally missed it when merging the conflicts earlier!

PS: also removed locales relating to the e-mail mode which will be phased out

LoneRifle commented on frontend/src/features/workspace/components/CreateFormModal/CreateFormModalContent/FormResponseOptions.tsx:
```diff
@@ -70,13 +73,9 @@ export const FormResponseOptions = forwardRef<
-              { text: 'Supports email submissions' },
-              { text: 'Supports Singpass & Myinfo' },
-              { text: 'Supports webhooks for integrations' },
-              {
-                text: 'Up to Confidential (Cloud-Eligible) and Sensitive (High) data',
-                badge: 'new',
-              },
```
This has gone missing, probably just because the PR has been around for a while. I'll try to add it later, unless you get to it first.

LoneRifle commented on frontend/src/i18n/locales/constants/validation/en-sg.ts:
nit - the `Error` suffix in many of the keys is somewhat redundant, given that most of the keys imply a validation error of some manner

LoneRifle reviewed (approved):
lgtm otherwise! I'll try to make the change I pointed out earlier, unless you're quicker than I am.

LoneRifle committed:
fix(i18n): update response options text
